README.md
strong-docker-build
===================


CHANGES.md
2016-05-06, Version 1.1.1
=========================

 * update copyright notices and license (Ryan Graham)

 * Refer to licenses with a link (Sam Roberts)

 * Use strongloop conventions for licensing (Sam Roberts)

 * test: enable coverage reporting (Ryan Graham)


2015-06-09, Version 1.1.0
=========================

 * test: fix misleading assertion (Ryan Graham)

 * Update license to dual Artistic-2.0/StrongLoop (Ryan Graham)

 * deps: update 0.x deps to latest (Ryan Graham)

 * load dockerode as lazily as possible (Ryan Graham)

 * deps: add lodash (Ryan Graham)


2015-05-21, Version 1.0.0
=========================

 * First release!


test/test-build.js
// Copyright IBM Corp. 2015. All Rights Reserved.
// Node module: strong-docker-build
// This file is licensed under the Artistic License 2.0.
// License text available at https://opensource.org/licenses/Artistic-2.0

var builder = require('../');
var docker = require('../lib/lazy-docker');
var path = require('path');
var tap = require('tap');

var appRoot = path.resolve(__dirname, 'sample-app');
var builtImage = null;

tap.test('node base image', {timeout: 90000}, function(t) {
  docker().pull('node:0.10', {repo: 'node'}, function(err, stream) {
    t.ifError(err, 'start pull node:0.10 without error');
    docker().modem.followProgress(stream, function(err) {
      t.ifError(err, 'pull node:0.10 without error');
      t.end();
    });
  });
});

tap.test('debian:jessie base image', {timeout: 90000}, function(t) {
  docker().pull('debian:jessie', {repo: 'debian'}, function(err, stream) {
    t.ifError(err, 'start pull debian:jessie without error');
    docker().modem.followProgress(stream, function(err) {
      t.ifError(err, 'finish pull debian:jessie without error');
      t.end();
    });
  });
});

tap.test('remove existing default', function(t) {
  var img = docker().getImage('sl-docker-run/test-app:0.0.0');
  img.remove({force: true}, function() {
    t.ok(true, 'removed existing image if existed');
    t.end();
  });
});

tap.test('build default', {timeout: 180000}, function(t) {
  builder.buildDeployImage({appRoot: appRoot}, function(err, res) {
    t.ifError(err, 'builds successfully');
    t.equal(res.name, 'sl-docker-run/test-app:0.0.0', 'Image named and tagged');
    t.end();
  });
});

tap.test('wait for default image', function(t) {
  t.ok(true, 'let docker finalize image');
  setTimeout(t.end.bind(t), 750);
});

tap.test('inspect image', function(t) {
  docker().getImage('sl-docker-run/test-app:0.0.0')
          .inspect(function(err, image) {
    t.ifError(err, 'Image should be inspectable');
    builtImage = image;
    var cfg = image.Config;
    var size = image.VirtualSize;
    var expectedEntryPoint = ['/usr/local/bin/sl-run', '--control', '8700'];
    t.equivalent(cfg.Entrypoint, expectedEntryPoint, 'Runs sl-run');
    t.equal(cfg.WorkingDir, '/app', 'Runs from /app');
    t.assert(size < 180 * 1024 * 1024, 'Image is less than 180MB');
    t.end();
  });
});

tap.test('remove existing custom image', function(t) {
  docker().getImage('strong-docker-build:test')
          .remove({force: true}, function() {
    t.ok(true, 'removed existing image if existed');
    t.end();
  });
});

tap.test('build custom image', {timeout: 180000}, function(t) {
  var opts = {appRoot: appRoot, imgName: 'strong-docker-build:test'};
  builder.buildDeployImage(opts, function(err, res) {
    t.ifError(err, 'rebuild builds successfully');
    t.equal(res.name, 'strong-docker-build:test', 'Rebuild named and tagged');
    t.end();
  });
});

tap.test('wait for  custom image', function(t) {
  t.ok(true, 'let docker finalize image');
  setTimeout(t.end.bind(t), 750);
});

tap.test('inspect rebuilt image', function(t) {
  docker().getImage('strong-docker-build:test').inspect(function(err, image) {
    t.ifError(err, 'Rebuild image should be inspectable');
    t.ok(image.VirtualSize * 0.99999 < builtImage.VirtualSize &&
         image.VirtualSize * 1.00001 > builtImage.VirtualSize,
         'Rebuild image within 0.0001% of original image size');
    t.end();
  });
});


test/sample-app/index.js
// Copyright IBM Corp. 2015. All Rights Reserved.
// Node module: strong-docker-build
// This file is licensed under the Artistic License 2.0.
// License text available at https://opensource.org/licenses/Artistic-2.0

var env = process.env;
var fs = require('fs');
var http = require('http');

console.log('PID %d', process.pid);
console.log('CWD %s', process.cwd());
console.log('ENV.supervisor_profile:', env.supervisor_profile);
console.log('ENV.PATH:', env.PATH);
console.log('ENV.PWD:', env.PWD);

// Check binary dependencies were compiled
require('buffertools');
// Check dev dependencies weren't installed
var debug = null;
try {
  debug = require('debug');
} catch(e) {
  console.log('debug missing, as expected');
}
if (debug) {
  throw Error('dev dependency "debug" was installed');
}

http.createServer(onRequest)
    .listen(process.env.PORT || 0, function() {
  console.log('pid %d listening on %s', process.pid, this.address().port);

  // Used to verify process existence/health
  fs.writeFileSync('app.pid', process.pid);
  fs.writeFileSync('app.port', this.address().port);
});

function onRequest(req, res) {
  switch (req.url) {
    case '/env':
      res.end(JSON.stringify(process.env, null, 2) + '\n\n');
      break;
    default:
      res.end(req.method + ' ' + req.url + '\n\n');
      break;
  }
}


lib/exec.js
// Copyright IBM Corp. 2015. All Rights Reserved.
// Node module: strong-docker-build
// This file is licensed under the Artistic License 2.0.
// License text available at https://opensource.org/licenses/Artistic-2.0

var async = require('async');
var docker = require('./lazy-docker');
var fnpm = require('fstream-npm');
var path = require('path');
var tar = require('tar');
var through = require('through');

exports.simple = simpleExec;
exports.streamIn = streamInExec;
exports.addApp = addAppExec;

function simpleExec(container, cmd, callback) {
  var execOpts = {
    AttachStdout: true,
    Cmd: cmd,
  };
  return async.waterfall([
    createExec,
    startExec,
    captureExec,
    waitExec,
  ], callback);

  function createExec(next) {
    container.exec(execOpts, next);
  }
}

function startExec(exec, next) {
  exec.start({stdin: true}, next);
}

function waitExec(stream, next) {
  stream.on('end', next);
  stream.on('error', next);
}

function streamInExec(container, cmd, callback) {
  var execOpts = {
    AttachStdout: true,
    AttachStderr: true,
    AttachStdin: true,
    Cmd: cmd,
  };

  return async.waterfall([
    createExec,
    startExec,
    captureExec,
  ], callback);

  function createExec(next) {
    container.exec(execOpts, next);
  }
}

function captureExec(stream, next) {
  docker().modem.demuxStream(stream, through(dot), process.stdout);
  stream.on('end', function() {
    if (dot.written) {
      console.log('done');
    }
  });

  return next(null, stream);

  function dot() {
    dot.written = true;
    process.stdout.write('.');
  }
}

function addAppExec(container, src, dst, callback) {
  var cmd = ['tar', '-C', dst, '--strip-components', '1', '-xvf-'];
  var pkg = {path: path.resolve(src), type: 'Directory'};

  // mimic 'npm pack'
  return async.waterfall([
    createSink,
    streamApp,
  ], callback);

  function createSink(next) {
    streamInExec(container, cmd, next);
  }

  function streamApp(stream, next) {
    fnpm(pkg).pipe(tar.Pack()).pipe(stream).on('end', next);
  }
}


lib/build.js
var async = require('async');
var fnpm = require('fstream-npm');
var fstream = require('fstream');
var path = require('path');
var util = require('./util');
var tar = require('tar');

exports.packageStream = packageStream;
exports.listStream = listStream;
exports.image = buildImage;

// mimic 'npm pack' as an fstream
function packageStream(pkgPath) {
  return fnpm({path: pkgPath, type: 'Directory', isDirectory: true});
}

function listStream(files) {
  var root = path.dirname(files[0]);
  return fstream.Reader({path: root, root: root, filter: inFiles});

  function inFiles() {
    return this.path === root || files.indexOf(this.path) !== -1;
  }
}

function buildImage(docker, dockerfile, imgName, inputStream, callback) {
  return async.waterfall([
    startBuild,
    dumpStream,
    util.success(imgName),
  ], callback);

  function startBuild(callback) {
    var buildContext = createBuildContext(inputStream);
    var buildOpts = {dockerfile: dockerfile, t: imgName};
    docker.buildImage(buildContext, buildOpts, callback);
  }

  function dumpStream(stream, next) {
    docker.modem.followProgress(stream, onFinish, log);

    function onFinish(err) {
      next(err); // drop output
    }

    function log(event) {
      if (event.stream) {
        process.stderr.write(event.stream);
      }
    }
  }

  function createBuildContext(inputStream) {
    var buildContext = tar.Pack();
    var dockerfiles = path.resolve(__dirname, 'dockerfiles');
    buildContext.add(fstream.Reader(dockerfiles));
    return inputStream.pipe(buildContext);
  }
}


lib/image.js
// Copyright IBM Corp. 2015. All Rights Reserved.
// Node module: strong-docker-build
// This file is licensed under the Artistic License 2.0.
// License text available at https://opensource.org/licenses/Artistic-2.0

var async = require('async');
var docker = require('./lazy-docker');
var through = require('through');

exports.singleLayer = singleLayerImage;
exports.copy = copyFromContainerImage;
exports.start = startImage;

// Execut a command on an image, returning a new image of the result
function singleLayerImage(img, cmd, callback) {
  var containerConfig = {
    Image: img,
    Entrypoint: cmd,
    Cmd: null,
  };
  var imageConfig = {
    comment: 'Built by strong-docker-build',
    author: 'strong-docker-build@' + require('../package.json').version,
  };
  var container = null;
  var image = null;

  return async.waterfall([
    create,
    start,
    wait,
    commit,
    cleanup,
  ], function(err) {
    callback(err, image);
  });

  function create(next) {
    docker().createContainer(containerConfig, next);
  }

  function start(newContainer, next) {
    container = newContainer;
    container.start({}, next);
  }

  function wait(stream, next) {
    container.wait(next);
  }

  function commit(res, next) {
    container.commit(imageConfig, next);
  }

  function cleanup(img, next) {
    image = img;
    container.remove(next);
  }
}

// create new container from base image and paths from another container
function copyFromContainerImage(srcContainer, srcPaths, baseImage, callback) {
  var destContainer = null;
  var tarc = {
    AttachStdout: true,
    AttachStderr: true,
    Cmd: [
      'tar', '-cf-', '-C', '/',
    ].concat(srcPaths),
  };
  var tarx = {
    AttachStdout: true,
    AttachStderr: true,
    AttachStdin: true,
    OpenStdin: true,
    StdinOnce: true,
    Entrypoint: ['tar', '-C', '/', '-xvpf-'],
    Cmd: [],
    Image: baseImage,
  };
  var files = 0;
  var fileCounter = through(function count(d) {
    files += d.toString().split('\n').length - 1;
  });
  var bytes = 0;
  var tarPipe = through(function inc(d) {
    bytes += d.length;
    this.queue(d);
  });

  return async.series([
    injectBuild,
    extractBuild,
    wait,
  ], returnResults);

  function injectBuild(next) {
    async.waterfall([
      createTarX,
      attachTarX,
      startTarX,
    ], next);
  }

  function extractBuild(next) {
    async.waterfall([
      createTarC,
      startTarC,
      captureTarC,
    ], next);
  }

  function createTarX(next) {
    docker().createContainer(tarx, next);
  }

  function attachTarX(c, next) {
    destContainer = c;
    var attachOpts = {
      stdin: true,
      stdout: true,
      stderr: true,
      stream: true,
    };
    destContainer.attach(attachOpts, next);
  }

  function startTarX(stream, next) {
    docker().modem.demuxStream(stream, fileCounter, process.stdout);
    tarPipe.pipe(stream.req);
    destContainer.start(tarx, next);
  }

  function createTarC(next) {
    srcContainer.exec(tarc, next);
  }

  function startTarC(exec, next) {
    exec.start(tarc, next);
  }

  function captureTarC(stream, next) {
    docker().modem.demuxStream(stream, tarPipe, process.stdout);
    stream.on('end', function() {
      tarPipe.end();
    });
    next();
  }

  function wait(next) {
    destContainer.wait(next);
  }

  function returnResults(err) {
    callback(err, destContainer, {bytes: bytes, files: files});
  }
}

// create a container from an image suitable for running multiple execs on
function startImage(baseImage, env, callback) {
  var container = null;
  return async.waterfall([
    create,
    start,
    returnContainer,
  ], callback);

  function create(next) {
    var opts = {
      Image: baseImage,
      Cmd: ['sleep', '1000'],
      Env: env,
    };
    docker().createContainer(opts, next);
  }

  function start(c, next) {
    container = c;
    c.start(next);
  }

  function returnContainer(thing, next) {
    next(null, container, thing);
  }
}


lib/extract.js
var async = require('async');
var fs = require('fs');
var path = require('path');
var success = require('./util').success;

exports.paths = extractPaths;

function extractPaths(docker, imgId, paths, callback) {
  var tempContainer = null;

  return async.waterfall([
    makeTempContainer,
    extractPathsFromContainer,
    cleanupTempContainer
  ], callback);

  function makeTempContainer(next) {
    docker.createContainer({Image: imgId}, next);
  }

  function extractPathsFromContainer(container, callback) {
    var absolutePaths = paths.map(resolvePath);
    tempContainer = container;
    async.map(absolutePaths, async.apply(extractPath, container), callback);
  }

  function cleanupTempContainer(paths, next) {
    tempContainer.remove(function(err) {
      next(err, paths);
    });
  }
}

function resolvePath(pathSpec) {
  pathSpec.tar = path.resolve(pathSpec.tar);
  return pathSpec;
}

function extractPath(container, pathSpec, callback) {
  return async.waterfall([
    getTarStream,
    writeFile,
    success(pathSpec.tar),
  ], callback);

  function getTarStream(callback) {
    container.copy({Resource: pathSpec.path}, callback);
  }

  function writeFile(stream, callback) {
    var file = fs.createWriteStream(pathSpec.tar);
    console.error('extracting %s -> %s', pathSpec.path, pathSpec.tar);
    stream.pipe(file).on('finish', callback);
  }
}


index.js
// Copyright IBM Corp. 2015. All Rights Reserved.
// Node module: strong-docker-build
// This file is licensed under the Artistic License 2.0.
// License text available at https://opensource.org/licenses/Artistic-2.0

var async = require('async');
var exec = require('./lib/exec');
var fmt = require('util').format;
var image = require('./lib/image');
var path = require('path');

exports.buildDeployImage = buildDeployImage;

function buildDeployImage(opts, callback) {
  var app = require(path.resolve(opts.appRoot, 'package.json'));
  var repo = extractRepoName(opts.imgName, 'sl-docker-run/' + app.name);
  var tag = extractTagName(opts.imgName, app.version);
  var nodeVersion = opts.nodeVersion || '0.10';
  var preDeployImage = null;
  var buildContainer = null;
  var deployContainer = null;
  var deployImage = null;
  var buildDetails = null;

  return async.series([
    createPreDeploy,
    createBuild,
    inspectBuild,
    copyBuildToDeploy,
    commitDeployContainer,
    cleanupBuild, cleanupDeploy,
  ], function(err) {
    callback(err, {name: repo + ':' + tag, id: deployImage});
  });

  function createBuild(next) {
    var container = null;
    var env = ['npm_config_spin=false'];
    if (process.env.npm_config_registry) {
      env.push('npm_config_registry=' + process.env.npm_config_registry);
    }
    var steps = [
      FROM(fmt('node:%s', nodeVersion)),
      RUN(['mkdir', '-p', '/app']),
      ADD(opts.appRoot, '/app'),
      RUN(['useradd', '-m', 'strongloop']),
      RUN(['chown', '-R', 'strongloop:strongloop', '/app', '/usr/local']),
      RUN(as('strongloop', 'cd /app && npm install --production')),
    ];

    var installStrongSupervisor = [
      RUN(as('strongloop', 'npm install -g strong-supervisor')),
    ];
    var installCustomSupervisor = [
      RUN(['mkdir', '-p', '/supervisor']),
      ADD(opts.supervisor, '/supervisor'),
      RUN(as('strongloop', 'npm install -g /supervisor')),
    ];

    if (opts.supervisor) {
      steps = steps.concat(installCustomSupervisor);
    } else {
      steps = steps.concat(installStrongSupervisor);
    }

    return async.series(steps, next);

    function FROM(baseImage) {
      return startAndCreate;

      function startAndCreate(next) {
        console.log('[build] FROM %s', baseImage);
        image.start(baseImage, env, function(err, c) {
          buildContainer = container = c;
          next(err);
        });
      }
    }

    function RUN(cmd) {
      return runCmd;

      function runCmd(next) {
        console.log('[build] RUN %s', cmd.join(' '));
        exec.simple(container, cmd, next);
      }
    }

    function ADD(src, dst) {
      return addApp;

      function addApp(next) {
        console.log('[build] ADD %s %s', path.relative('.', src), dst);
        exec.addApp(container, src, dst, next);
      }
    }

    function as(user, cmd) {
      return ['su', user, '-c', cmd];
    }
  }

  function inspectBuild(next) {
    buildContainer.inspect(function(err, details) {
      buildDetails = details;
      next(err);
    });
  }

  function createPreDeploy(next) {
    var baseImage = 'debian:jessie';
    var cmd = ['useradd', '-m', 'strongloop'];
    image.singleLayer(baseImage, cmd, function(err, res) {
      preDeployImage = res && res.Id;
      next(err);
    });
  }

  function copyBuildToDeploy(next) {
    var paths = [
      'app',
      'usr/local/bin/node',
      'usr/local/bin/sl-run',
      'usr/local/lib/node_modules/strong-supervisor',
    ];
    console.log('Copying build results from build container to final image');
    image.copy(buildContainer, paths, preDeployImage, function(err, c, counts) {
      console.log('Bytes read: %d', counts && counts.bytes);
      console.log('Files written: %d', counts && counts.files);
      deployContainer = c;
      next(err);
    });
  }

  function commitDeployContainer(next) {
    var imgConfig = {
      Image: 'debian:jessie',
      User: 'strongloop',
      WorkingDir: '/app',
      Env: [
        'PORT=3000',
        'STRONGLOOP_CLUSTER=CPU',
      ].concat(buildDetails.Config.Env.filter(function(e) {
        return /NODE_VERSION/.test(e);
      })),
      ExposedPorts: {
        '8700/tcp': {},
        '3000/tcp': {},
      },
      Entrypoint: [
        '/usr/local/bin/sl-run', '--control', '8700',
      ],
      repo: repo,
      tag: tag,
      comment: 'Built by strong-docker-build',
      author: 'strong-docker-build@' + require('./package.json').version,
    };
    console.log('[deploy] commiting: %s as %s:%s',
                deployContainer.id.slice(0, 12),
                imgConfig.repo, imgConfig.tag);
    deployContainer.commit(imgConfig, function(err, res) {
      deployImage = res && res.Id;
      next(err);
    });
  }

  function cleanupBuild(next) {
    console.log('[build] removing container');
    buildContainer.remove({v: true, force: true}, next);
  }

  function cleanupDeploy(next) {
    console.log('[deploy] removing container');
    deployContainer.remove({v: true, force: true}, next);
  }
}

function extractRepoName(imgName, dflt) {
  var repo = imgName && imgName.split(':')[0];
  return repo || dflt;
}

function extractTagName(imgName, dflt) {
  var tag = imgName && imgName.split(':')[1];
  return tag || dflt;
}
